src/kemal/exception_handler.cr

```diff
@@ -11,12 +11,44 @@ module Kemal
     rescue ex : Kemal::Exceptions::CustomException
       call_exception_with_status_code(context, ex, context.response.status_code)
     rescue ex : Exception
+      # Use error handler defined for the current exception if it exists
+      return call_exception_with_exception(context, ex, 500) if Kemal.config.error_handlers.has_key?(ex.class)
+
+      # Use error handler for an ancestor of the current exception if it exists
+      Kemal.config.error_handlers.each_key do |key|
+        if key.is_a? Exception.class && ex.class <= key
+          return call_exception_with_exception(context, ex, 500, override_handler_used: key)
+        end
+      end
+
       log("Exception: #{ex.inspect_with_backtrace}")
+      # Else use generic 500 handler if defined
       return call_exception_with_status_code(context, ex, 500) if Kemal.config.error_handlers.has_key?(500)
       verbosity = Kemal.config.env == "production" ? false : true
       render_500(context, ex, verbosity)
     end
 
+    # Calls the defined error handler for the given exception if it exists
+    #
+    # By default it tries to use the handler that is defined for the given exception. However, another
+    # handler can be used via the `override_handler_used` parameter.
+    private def call_exception_with_exception(context : HTTP::Server::Context, exception : Exception, status_code : Int32 = 500, override_handler_used : (Exception.class)? = nil)
+      return if context.response.closed?
+
+      if !override_handler_used
+        handler_to_use = exception.class
+      else
+        handler_to_use = override_handler_used
+      end
+
+      if !Kemal.config.error_handlers.empty? && Kemal.config.error_handlers.has_key?(handler_to_use)
```

Comment:
**performance:** The combination of `h.has_key?(key)` and later `h[key]` is inefficient because it requires two hash lookups for the same key. Instead, you could directly query the value with `h[key]?` and use that as a condition. A `nil` return value would indicate the key does not exist.
Even better, you could already retrieve the handler while matching the exception (use `each` instead of `each_key` and you get the value for free) and pass it directly to this method. That removes two unnecessary hash lookups.